### PR TITLE
Number input widget

### PR DESCRIPTION
**Pull request by hangleang**
Hi there, I would like to PR that implement number input widget in #209 issue. I have also added a simple example for show case this widget. I'm look forward to see any idea, feel free to improve it or leave a comment here.

### CODE REVIEW

File: examples/number_input/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+};
+
+#[derive(Default)]
+pub struct NumberInputDemo {
+    state: number_input::State,
+    value: u8,
+}
+
+#[derive(Debug, Clone)]
+pub enum NumInpMessage {
```
Comment (13r0ck) [inline examples/number_input/src/main.rs:13]: pub enum Message

File: examples/number_input/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        window: window::Settings {
+            size: (250, 200),
+            ..Default::default()
+        },
+        ..Settings::default()
+    })
+}
+
+impl Sandbox for NumberInputDemo {
+    type Message = NumInpMessage;
```
Comment (13r0ck) [inline examples/number_input/src/main.rs:29]: type Message - Message;

File: examples/number_input/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            value: 27,
+            ..Self::default()
+        }
+    }
+
+    fn title(&self) -> String {
+        String::from("Number Input Demo")
+    }
+
+    fn update(&mut self, message: Self::Message) {
```
Comment (13r0ck) [inline examples/number_input/src/main.rs:42]: fn update(&mut self, message: Message)

File: examples/number_input/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        match message {
+            Self::Message::NumInpChanged(val) => {
```
Comment (13r0ck) [inline examples/number_input/src/main.rs:44]: Message::NumInpChanged

File: examples/number_input/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                self.value = val;
+            }
+        }
+    }
+
+    fn view(&mut self) -> Element<Self::Message> {
```
Comment (13r0ck) [inline examples/number_input/src/main.rs:50]: fn view(&mut self) -> Element<Message>

**Comment by 13r0ck** [comment]
It looks good so far, there are some comments about how to remove the `Self` from your messages. I really wish that it accepted scroll inputs to increase and decrease the values, as well if I hold on the up/down arrow it would make sense for the value to increase while held. 

There is a bug with your implementation of the max value. Lets say I clear the input to `0` then i put my cursor infront of the `|0` (like that) then i press, for example 9, then the input is `9|0` with my cursor in the middle. Though if I press 9 again the input is blocked rather than overwriting the 0. If would feel more natural to go from `9|0` -> `99|` which is still under 255 (the max in your example), though your code checks the input as `99|0` which is above 255 and blocks the input. Hopefully that makes sense?

**Review by hecrj — COMMENTED** [review]
Thank you so much for this! :smile: 

It may be interesting to contribute this widget to [`iced_aw`] first. This is, since very recently, the official community crate for additional widgets (see #761). I think it will be easier for users to contribute and improve the widget further there and then, once it eventually matures enough, we could migrate it to this repository.

Ideally, we should leave this repository for important fixes and core changes. This way, I can stop bottlenecking interesting contributions while focusing on the core parts of the library!

[`iced_aw`]: https://github.com/iced-rs/iced_aw

**Comment by Kaiden42** [comment]
> It may be interesting to contribute this widget to iced_aw first.

Sounds like a good idea 😉. If you have problems with moving your widget to iced_aw, please let me know.

>  I really wish that it accepted scroll inputs to increase and decrease the values,

This can be easily done with `mouse::Event::WheelScrolled { delta }`. You can take a look at the [Color Picker](https://github.com/iced-rs/iced_aw/blob/main/src/native/overlay/color_picker.rs#L273) implementation for this.

> as well if I hold on the up/down arrow it would make sense for the value to increase while held.

As far as I know, this is currently not possible. It would require the widget to send an update/message by itself after some period, wich is currently not implemented. This is slightly related to #31 and #560 .

> There is a bug with your implementation of the max value. Lets say I clear the input to 0 then i put my cursor infront of the |0 (like that) then i press, for example 9, then the input is 9|0 with my cursor in the middle. Though if I press 9 again the input is blocked rather than overwriting the 0. If would feel more natural to go from 9|0 -> 99| which is still under 255 (the max in your example), though your code checks the input as 99|0 which is above 255 and blocks the input. Hopefully I makes sense?

For comparison, I looked at how the number-input field of [HTML](https://wiki.selfhtml.org/wiki/HTML/Tutorials/Formulare/input/number) handles this (chromium). The maximum value there, is not considered at all with the text input. With a max value of 220 you can easily enter 9999990 or similar. I find persöhnlich also not just intuitive.

However, with your suggestion I would rather go for `99|0` -> 255 instead of `99|0` -> 99. Maybe we can look at how other GUI libraries handle this case and pick out the best behavior for us.

**Comment by hangleang** [comment]
> It looks good so far, there are some comments about how to remove the `Self` from your messages. I really wish that it accepted scroll inputs to increase and decrease the values, as well if I hold on the up/down arrow it would make sense for the value to increase while held.

Thank for review @13r0ck. I've format the example for convenient as you told. I also merge with your branch(more_scroll) and in the integration, I've change on scroll event to increase or decrease the value directly inside widget logic because I don't want user to handle those two messages by themselves, but the logic its self is still the same. I was given you credit on that section, Hope you don't mind on that.

> There is a bug with your implementation of the max value. Lets say I clear the input to `0` then i put my cursor infront of the `|0` (like that) then i press, for example 9, then the input is `9|0` with my cursor in the middle. Though if I press 9 again the input is blocked rather than overwriting the 0. If would feel more natural to go from `9|0` -> `99|` which is still under 255 (the max in your example), though your code checks the input as `99|0` which is above 255 and blocks the input. Hopefully I makes sense?

This update is fixed those bugs if I do understand you well. So, please check on that again if there is some point I missing. Leave it a comment. Thank again for spending your time on review.

**Comment by hangleang** [comment]
> Thank you so much for this! smile
> It may be interesting to contribute this widget to [`iced_aw`](https://github.com/iced-rs/iced_aw) first. This is, since very recently, the official community crate for additional widgets (see #761). I think it will be easier for users to contribute and improve the widget further there and then, once it eventually matures enough, we could migrate it to this repository.

Yeah, you are welcome. I'll put this widget in that repo when the time allowed. I got a lot of ideas from the repo, hahaa.

> Ideally, we should leave this repository for important fixes and core changes. This way, I can stop bottlenecking interesting contributions while focusing on the core parts of the library!

Ohh I see. Thank for warm welcome.

**Comment by hangleang** [comment]
> > It may be interesting to contribute this widget to iced_aw first.

Thank for interested on the widget. Good if we can works together.

> Sounds like a good idea wink. If you have problems with moving your widget to iced_aw, please let me know.

Not now, but soon hahaa. 

> > I really wish that it accepted scroll inputs to increase and decrease the values,
> This can be easily done with `mouse::Event::WheelScrolled { delta }`. You can take a look at the [Color Picker](https://github.com/iced-rs/iced_aw/blob/main/src/native/overlay/color_picker.rs#L273) implementation for this.

This update has implementation on that event which I merge from @13r0ck. Thank to @13r0ck.

> > There is a bug with your implementation of the max value. Lets say I clear the input to 0 then i put my cursor infront of the |0 (like that) then i press, for example 9, then the input is 9|0 with my cursor in the middle. Though if I press 9 again the input is blocked rather than overwriting the 0. If would feel more natural to go from 9|0 -> 99| which is still under 255 (the max in your example), though your code checks the input as 99|0 which is above 255 and blocks the input. Hopefully I makes sense?
> 
> For comparison, I looked at how the number-input field of [HTML](https://wiki.selfhtml.org/wiki/HTML/Tutorials/Formulare/input/number) handles this (chromium). The maximum value there, is not considered at all with the text input. With a max value of 220 you can easily enter 9999990 or similar. I find persöhnlich also not just intuitive.
> 
> However, with your suggestion I would rather go for `99|0` -> 255 instead of `99|0` -> 99. Maybe we can look at how other GUI libraries handle this case and pick out the best behavior for us.

This widget is required max value for checking valid number, not like the number input in HTML.

**Comment by 13r0ck** [comment]
I am going to hold out on more input until this PR is moved over to [Iced aw](https://github.com/iced-rs/iced_aw). This is going to be a useful widget!

**Comment by Kaiden42** [comment]
Since this is now merged into `iced_aw` this PR can be closed now. :slightly_smiling_face:

**Pull request closed without merging**